File: 14501.c
#include <stdio.h>
int max(int a,int b){
    return a>b ? a:b;
}

int main(){
    int d[100000];
    int n=0;
    scanf("%d",&n);
    int t[n+1];
    int p[n+1];

    for(int i=1;i<=n;i++){
        scanf("%d %d",&t[i],&p[i]);
    }

    for(int j=1;j<=n;j++){
        if(j+t[j]<=n){
            d[j+t[j]] = max(d[j+t[j]],d[j]+p[j]);
        }
    }

}

File: 14501.py

N = int(input())

g = []
dp = [0 for _ in range(N+1)]
day = 0
for i in range(N):
    t, p = map(int, input().split())
    g.append([i, t, p])

# 뒤에서부터 DP로 풀이 -> 가능한 일수가 있을 때만 pi ++
for i in range(N-1, -1, -1):
        if g[i][1] + i <= N:
            dp[i] = max(dp[i+1], g[i][2]+dp[i + g[i][1]])
        else:
            dp[i] = dp[i+1]

print(dp[0])

